src/graph/executor/query/ExpandAllExecutor.cpp

```diff
@@ -0,0 +1,379 @@
+  auto iter = std::make_unique<GetNeighborsIter>(listVal);
+  if (iter->numRows() == 0) {
+    return Status::OK();
+  }
+  auto size = iter->size();
+  std::vector<int64_t> samples;
+  if (sample_) {
+    algorithm::ReservoirSampling<int64_t> sampler(curMaxLimit_, size);
+    samples = sampler.samples();
+    std::sort(samples.begin(), samples.end(), [](int64_t a, int64_t b) { return a > b; });
+  }
+
+  QueryExpressionContext ctx(ectx_);
+  std::unordered_map<Value, std::unordered_set<Value>> dst2VidsMap;
+  std::unordered_set<Value> visitedVids;
+  std::vector<List> adjEdgeProps;
+  List curVertexProps;
+  Value curVid;
+  if (iter->valid() && vertexColumns_) {
+    for (auto& col : vertexColumns_->columns()) {
+      Value val = col->expr()->eval(ctx(iter.get()));
+      curVertexProps.values.emplace_back(std::move(val));
+    }
+  }
+  for (; iter->valid(); iter->next()) {
+    List edgeProps;
+    if (edgeColumns_) {
+      for (auto& col : edgeColumns_->columns()) {
+        Value val = col->expr()->eval(ctx(iter.get()));
+        edgeProps.values.emplace_back(std::move(val));
+      }
+    }
+    // get the start vid of the next step through dst when expanding in the cache
+    auto dst = iter->getEdgeProp("*", nebula::kDst);
+    if (dst.isNull()) {
+      // (TEST)
+      continue;
+    }
+    edgeProps.values.emplace_back(dst);
+    const auto& vid = iter->getColumn(0);
+    curVid = curVid.empty() ? vid : curVid;
+    if (curVid != vid) {
+      adjEdgeProps.emplace_back(std::move(curVertexProps));
+      adjList_.emplace(curVid, std::move(adjEdgeProps));
+      curVid = vid;
+      if (vertexColumns_) {
+        for (auto& col : vertexColumns_->columns()) {
+          Value val = col->expr()->eval(ctx(iter.get()));
+          curVertexProps.values.emplace_back(std::move(val));
+        }
+      }
+    }
+    adjEdgeProps.emplace_back(edgeProps);
+
+    if (sample_) {
+      if (samples.empty()) {
+        continue;
+      }
+      if (curLimit_++ != samples.back()) {
+        continue;
+      } else {
+        samples.pop_back();
+      }
+    } else {
+      if (curLimit_++ >= curMaxLimit_) {
+        continue;
+      }
+    }
+
+    if (joinInput_) {
+      auto findVid = preDst2VidsMap_.find(vid);
+      buildResult(findVid->second, curVertexProps, edgeProps);
+      updateDst2VidsMap(dst2VidsMap, vid, dst);
+    } else {
+      buildResult(curVertexProps, edgeProps);
+    }
+
+    if (adjList_.find(dst) == adjList_.end()) {
+      nextStepVids_.emplace(dst);
+    } else {
+      visitedVids.emplace(dst);
+    }
+  }
+  if (!curVid.empty()) {
+    adjEdgeProps.emplace_back(std::move(curVertexProps));
+    adjList_.emplace(curVid, std::move(adjEdgeProps));
+  }
+  auto vidIter = nextStepVids_.begin();
+  while (vidIter != nextStepVids_.end()) {
+    if (adjList_.find(*vidIter) != adjList_.end()) {
+      visitedVids.emplace(*vidIter);
+      vidIter = nextStepVids_.erase(vidIter);
+    } else {
+      vidIter++;
+    }
+  }
```

Comment:
could you extract the same logics into the functions for the usage in `handleResponse` and `getNeighborsFromCache`?

Reply:
done